linera-chain/src/manager.rs

```diff
@@ -424,51 +426,60 @@ where
         key_pair: Option<&KeyPair>,
         local_time: Timestamp,
         blobs: BTreeMap<BlobId, Blob>,
-    ) -> Result<Option<ValidatedOrConfirmedVote>, ViewError> {
+    ) -> Result<Option<ValidatedOrConfirmedVote>, ChainError> {
         let round = proposal.content.round;
+        if key_pair.is_some() && round < self.current_round() {
+            return Ok(None);
+        }
 
         // If the validated block certificate is more recent, update our locked block.
         if let Some(lite_cert) = &proposal.validated_block_certificate {
             if self
                 .locked
                 .get()
                 .as_ref()
-                .map_or(true, |locked| locked.round < lite_cert.round)
+                .map_or(true, |locked| locked.round() < lite_cert.round)
             {
                 let value = Hashed::new(ValidatedBlock::new(executed_block.clone()));
                 if let Some(certificate) = lite_cert.clone().with_value(value) {
-                    self.set_locked(certificate, blobs)?;
+                    self.set_locked(LockedBlock::Regular(certificate), blobs)?;
                 }
             }
+        } else if round.is_fast() {
+            // The fast block also counts as locked.
+            self.set_locked(LockedBlock::Fast(proposal.clone()), blobs)?;
         }
 
+        let Some(key_pair) = key_pair else {
+            // Record the proposed block, in case it affects the current round number.
+            self.set_proposed(proposal);
+            self.update_current_round(local_time);
+            return Ok(None);
+        };
+
+        self.check_proposal_round(&proposal)?;
         // Record the proposed block, so it can be supplied to clients that request it.
-        self.proposed.set(Some(proposal));
+        self.set_proposed(proposal);
```

Comment:
Can we move this (and line below) to line 452?

Reply:
No, we can't update the current round before checking against the current round.

Reply:
Yes, this was conditional on the acceptance of the previous comment.